### DISCUSSION TITLE
Pouvoir générer des PDFs accessibles

### DISCUSSION THREAD

**Discussion by goulvench** [betagouv/cap#28] [Besoin]
### Besoin

De très nombreuses SE génèrent des PDFs, en s'appuyant sur différents outils : des librairies bas niveau, d'autres capables de convertir du HTML, voire des outils distribués sous forme d'images Docker… Mais aucun de ces outils ne génère de PDF accessible.

Les services de l'État ont l'obligation d'être accessibles, il serait donc logique que les PDFs générés le soient aussi.

Une instance partagée, interrogeable par une URL, permettrait à toutes les SE de générer des PDFs accessibles en s'épargnant la configuration d'un outil ad hoc. Des modèles d'intégration pourraient être partagés, et une base de styles issus du DSFR pourrait être proposée par défaut (police, couleurs).

### Solutions envisageables

1. [OpenHTML2PDF (Java)](https://github.com/danfickle/openhtmltopdf)
2. [Gotenberg (Go)](https://gotenberg.dev/)
3. [PDFme (Typescript)](https://github.com/pdfme/pdfme)
4. [React-PDF (React)](https://react-pdf.org/)
5. [WeasyPrint (Python)](https://doc.courtbouillon.org/weasyprint/stable/)
6. [Note-ministre (LaTeX)](https://github.com/arnaud-robin/note-ministre/)
7. [FerrumPDF](https://github.com/excid3/ferrum_pdf)

### Informations complémentaires

- [X] J'avais déjà entendu ce besoin au sein de la communauté beta.gouv
- [ ] Je me porte volontaire pour participer à la mise en place de la future solution choisie

**Comment by freesteph** [comment]
J'avais un début de solution sur https://github.com/freesteph/paperasse

EDIT: je répare le proto et j'update le README

**Comment by goulvench** [comment]
Ah oui, je n’avais pas en tête qu’Hexapdf peut générer des PDFs
accessibles. Si c’est vrai, c’est bon à savoir !

Attention à la licence, elle est un peu différente de la plupart des
gemmes.

**Comment by freesteph** [comment]
Je signale aussi l'excellent https://github.com/arnaud-robin/note-ministre/, qui est plus avancé mais potentiellement un peu plus dur à maintenir (tex+pandoc)

**Comment by weblaetitia** [comment]
pdfkit est une lib bas niveau, mais propose des options d'accessibilité : https://pdfkit.org/docs/accessibility.html

**Comment by goulvench** [comment]
Erwan m'a également signalé :
- [iText (Java)](https://itextpdf.com/products/convert-html-css-to-pdf-pdfhtml), intègre une option HTML->PDF accessible
- il est possible de générer des PDFs en ligne de commande à partir d'un modèle `odt`
- et que plusieurs services de l'état ont déjà mis des solutions en place.

**Reply by Erwan-G-le-Gall** [reply]
Merci Goulven.
Je précise : le format **.fodt** (Flat XML ODT OpenDocument) crée un unique document XML qui contient p. ex. les images sous forme de data-URI. Il peut être utilisé comme template. Et il est possible de lancer LibreOffice en mode serveur (instance headless permanente) pour réaliser la [conversion en CLI](https://help.libreoffice.org/latest/fr/text/shared/guide/convertfilters.html), option --convert-to. NB pour des volumétries raisonnables.
Apache **PDFbox** pour celles et ceux qui aiment Java (s’il y en a ;) ), très complet, bas niveau.
Pour iText si vous pouvez vous accommoder de la licence on a un [PoC](https://github.com/DISIC/generate-pdf-example) en HTMLtoPDF et en mode code pur.

**Reply by yaaax** [reply]
Salut !

Avec **iText** j’ai fait cet exemple [Generate PDF example (répo GitHub)](https://github.com/DISIC/generate-pdf-example).
[Chercher a11y](https://github.com/search?q=repo%3ADISIC%2Fgenerate-pdf-example%20a11y&type=code) pour voir les lignes de code importantes qui participent à exporter les bonnes balises pour l’accessibilité.

Note : je ne suis pas un "dev Java", le code n’est sûrement pas le plus beau. Mais ça donne des pistes intéressantes.

**Reply by Erwan-G-le-Gall** [reply]
@yaaax c’est le PoC évoqué :-D

**Comment by Erwan-G-le-Gall** [comment]
en LaTeX c’est possible avec la lib **tagpdf** https://www.latex-project.org/news/2024/03/27/PDFUA2-examples/
cf https://www.latex-project.org/news/2024/03/27/PDFUA2-examples/
ping @freesteph si _Note ministre_ veut [s’améliorer](https://github.com/arnaud-robin/note-ministre/issues/2) ;)

**Reply by freesteph** [reply]
faut pinger son auteur @arnaud-robin (qui vit pas loin de toi au bureau si mes souvenirs sont bons). Moi je suis team Paperpasse 😉 🍂

**Reply by freesteph** [reply]
cela dit je viens de regarder HexaPDF (qui est derrière Papepasse) et pas sûr de comprendre où en est le support de UA/2 (qui est spec de test de PDF ? une spec de PDF ? je suis voué à ne jamais comprendre le monde du PDF) 

https://hexapdf.gettalong.org/documentation/implementation-status.html
https://hexapdf.gettalong.org/documentation/pdfa/#best-practices

**Reply by Erwan-le-Gall** [reply]
@freesteph PDF/UA-1 est une spec de PDF (qui ajoute des contraintes) et [Matterhorn Protocol](https://pdfa.org/resource/the-matterhorn-protocol/) la spec de test correspondante

**Comment by Erwan-G-le-Gall** [comment]
c’est l’heure du cours gratuit :
PDF/UA Universal Accessibility sous-ensemble du PDF qui respecte certains points (avoir un titre dans les métadonnées du fichier…)
PDF/A-* Archives
PDF/A participe à l’accessibilité parce que pour l’archivage tu embarques les polices donc tu n’as pas de problèmes de substitution de police à l’arrache et les personnes dyslexiques te disent merci. Et le prérequis du balisage est assuré aussi.

**Comment by Erwan-G-le-Gall** [comment]
décortiquer hexapdf attendra que j’ai un vrai ordinateur sous la main, sous Linux (donc à la maison)

**Comment by revolunet** [comment]
on crack ça les 11/12 ? https://www.linkedin.com/feed/update/urn:li:activity:7285402608104534018/

**Comment by goulvench** [comment]
Je viens de voir que [Chrome génère des PDFs tagués](https://blog.chromium.org/2020/07/using-chrome-to-generate-more.html).
Couplé à [Ferrum PDF](https://github.com/excid3/ferrum_pdf), on pourrait créer une API qui prend une URL (ou du HTML) et renvoie un PDF (ou une image).

**Reply by Erwan-le-Gall** [reply]
Merci Goulven. Point d’attention sur les aspects qui ne relèvent pas de l’outil (contraste, pertinence des textes équivalents aux images…) mais du créateur du contenu (qu’il soit contributeur ou codeur :) ).
Par ailleurs le format EPUB3 est plutôt plus accessible et plus simple à créer (mais un peu méconnu).

**Comment by revolunet** [comment]
un service de generation pdf à base de puppeteer chez le GIP inclusion : https://github.com/gip-inclusion/pdf-generator
